Fix `#previously_new_record?` on destroyed records

adrianna-chang-shopify opened the pull request:
### Motivation / Background

Ref: #48794
cc @eizengan

Based on the [documentation](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-previously_new_record-3F) for `#previously_new_record?`:
> Returns true if this object was just created – that is, prior to the last save, the object didn't exist in the database and new_record? would have returned true.

IMO a record that was previously created and has since been destroyed and removed from the database should no longer be considered "just created", although one could argue that a `DELETE` is different from a "save".

### Detail

Set `@previously_new_record` to `false` in `#destroy` and `#delete` methods.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

adrianna-chang-shopify committed:
Fix #previously_new_record? on destroyed records

Ref: #48794

`#previously_new_record?` should return false for records that are
created and then destroyed.

eizengan commented:
> IMO a record that was previously created and has since been destroyed and removed from the database should no longer be considered "just created", although one could argue that a DELETE is different from a "save".

My take as well. 

> prior to the last save, the object didn't exist in the database and new_record? would have returned true

feels less useful than e.g.

> that is, prior to the _last call which modified that record_, the object didn't exist in the database and new_record? would have returned true

nvasilevski commented on activerecord/test/cases/base_test.rb:
```diff
@@ -968,6 +968,14 @@ def test_previously_new_record_returns_boolean
 
+  def test_previously_new_record_on_destroyed_record
+    topic = Topic.create
+    assert topic.previously_new_record?
```
Nitpick but generally `assert_predicate` provides a better error message. Even though we may never see this test fail in case if it fails `assert_predicate` will produce something like "expected `Topic<id:123>` to be :previously_new_record?" which allows for better understanding of the failure while `assert` will produce a very dry "expected false to be truthy" message 

```suggestion
    assert_predicate topic, :previously_new_record?
```

adrianna-chang-shopify replied:
Agreed, thanks Nikita!

nvasilevski commented on activerecord/test/cases/base_test.rb:
```diff
@@ -968,6 +968,14 @@ def test_previously_new_record_returns_boolean
+    assert topic.previously_new_record?
+
+    topic.destroy
+    assert_not topic.previously_new_record?
```
Even though I don't find `assert_not_predicate` name particularly friendly I think it pays off with a better error message :) 
```suggestion
    assert_not_predicate topic, :previously_new_record?
```

nvasilevski reviewed (approved):
Makes sense 👍

eileencodes reviewed (changes requested):
I thought about this for a bit and I'm kind of on the fence - it would depend whether we think a save and delete are different for this context. I think it is since a double save makes `previously_new_record?` return `false`. I think the docs should be updated though to reflect this change. Also with the slight change in behavior that might be surprising, can you add a CHANGELOG entry too?

adrianna-chang-shopify commented:
> it would depend whether we think a save and delete are different for this context.

Yeah, I was on the fence about this too. I ended up leaning towards these changes because the primary use case around `#previously_new_record?` seems to be for triggering callbacks on records that were newly created. Even though "save" and "delete" may be considered different in this context, we'd want to prevent a callback from firing off on the first "create" and then again on a destroy, as @eizengan demonstrated in the example in his issue.

I've updated the docs and added a CHANGELOG entry -- let me know what you think 😄
